pbs/pbsrequest.go

```diff
@@ -130,85 +122,82 @@ func ParsePBSRequest(r *http.Request, cache cache.Cache) (*PBSRequest, error) {
-	}
+		pbsReq.Device.UA = r.Header.Get("User-Agent")
 
-	if strings.Index(pbsReq.Url, "http") == -1 {
-		pbsReq.Url = fmt.Sprintf("http://%s", pbsReq.Url)
-	}
+		if r.Header.Get("X-Real-Ip") != "" {
+			pbsReq.Device.IP = r.Header.Get("X-Real-Ip")
+		} else {
+			ip, _, uerr := net.SplitHostPort(r.RemoteAddr)
+			if uerr == nil && net.ParseIP(ip) != nil {
+				pbsReq.Device.IP = ip
+			}
+		}
+		pbsReq.Url = r.Header.Get("Referer") // must be specified in the header
+		// TODO: this should explicitly put us in test mode
+		if r.FormValue("url_override") != "" {
+			pbsReq.Url = r.FormValue("url_override")
+		}
+		if strings.Index(pbsReq.Url, "http") == -1 {
+			pbsReq.Url = fmt.Sprintf("http://%s", pbsReq.Url)
+		}
 
-	url, err := url.Parse(pbsReq.Url)
-	if err != nil {
-		return nil, fmt.Errorf("Invalid URL '%s': %v", pbsReq.Url, err)
-	}
+		url, err := url.Parse(pbsReq.Url)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid URL '%s': %v", pbsReq.Url, err)
+		}
 
-	if url.Host == "" {
-		return nil, fmt.Errorf("Host not found from URL '%v'", url)
-	}
+		if url.Host == "" {
+			return nil, fmt.Errorf("Host not found from URL '%v'", url)
+		}
 
-	pbsReq.Domain, err = publicsuffix.EffectiveTLDPlusOne(url.Host)
-	if err != nil {
-		return nil, fmt.Errorf("Invalid URL '%s': %v", url.Host, err)
-	}
+		pbsReq.Domain, err = publicsuffix.EffectiveTLDPlusOne(url.Host)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid URL '%s': %v", url.Host, err)
+		}
 
-	_, err = cache.GetDomain(pbsReq.Domain)
-	if err != nil {
-		return nil, fmt.Errorf("Invalid URL %s", pbsReq.Domain)
+		// all domains must be in the whitelist
+		_, err = cache.GetDomain(pbsReq.Domain)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid URL %s", pbsReq.Domain)
+		}
+	} else {
+		/* should we whitelist bundles? Skipping for now
+		_, err = cache.GetDomain(pbsReq.App.Bundle)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid app bundle %s", pbsReq.App.Bundle)
+		}
+		*/
 	}
 
 	if r.FormValue("debug") == "1" {
 		pbsReq.IsDebug = true
 	}
 
-	// TODO: X-Forwarded-For
-	if r.Header.Get("X-Real-Ip") != "" {
-		pbsReq.IPAddress = r.Header.Get("X-Real-Ip")
-	} else {
-		ip, _, uerr := net.SplitHostPort(r.RemoteAddr)
-		if uerr == nil && net.ParseIP(ip) != nil {
-			pbsReq.IPAddress = ip
-		}
-	}
-
-	if r.Header.Get("X-Forwarded-Proto") != "" {
-		pbsReq.Protocol = r.Header.Get("X-Forwarded-Proto")
-	} else {
-		if r.TLS == nil {
-			pbsReq.Protocol = "http"
-		} else {
-			pbsReq.Protocol = "https"
+	if pbsReq.Secure == 0 {
+		if r.Header.Get("X-Forwarded-Proto") != "" {
+			if r.Header.Get("X-Forwarded-Proto") == "https" {
+				pbsReq.Secure = 1
+			}
+		} else if r.TLS != nil {
+			pbsReq.Secure = 1
 		}
 	}
-	if r.Header.Get("X-Forwarded-Host") != "" {
-		pbsReq.Hostname = r.Header.Get("X-Forwarded-Host")
-	} else {
-		pbsReq.Hostname = r.Host
-	}
 
-	pbsReq.ServerURL = fmt.Sprintf("%s://%s", pbsReq.Protocol, pbsReq.Hostname)
+	pbsReq.Bidders = make([]*PBSBidder, 0)
```

Comment:
That's length, not capacity (which would be the third argument). Right now you could technically have more PBSBidders than the number of adapters we know of (Index is still using a multi-bidder hack instead of the more elegant way Facebook is implemented; that's on my todo list). I'm going to add a constant for initial size here so we don't to do a dynamic lookup on number of adapters.